include/ignition/gazebo/comms/ICommsModel.hh

```diff
@@ -0,0 +1,202 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef IGNITION_GAZEBO_ICOMMSMODEL_HH_
+#define IGNITION_GAZEBO_ICOMMSMODEL_HH_
+
+#include <algorithm>
+#include <memory>
+
+#include <sdf/sdf.hh>
+#include <ignition/common/Profiler.hh>
+#include "ignition/gazebo/comms/Broker.hh"
+#include "ignition/gazebo/config.hh"
+#include "ignition/gazebo/EntityComponentManager.hh"
+#include "ignition/gazebo/System.hh"
+
+namespace ignition
+{
+namespace gazebo
+{
+// Inline bracket to help doxygen filtering.
+inline namespace IGNITION_GAZEBO_VERSION_NAMESPACE {
+namespace comms
+{
+  // Forward declarations.
+  class MsgManager;
+
+  /// \brief Abstract interface to define how the environment should handle
+  /// communication simulation. As an example, this class could be responsible
+  /// for handling dropouts, decay and packet collisions.
+  ///
+  /// The derived comms models can be configured with the following SDF
+  /// parameters:
+  ///
+  /// * Optional parameters:
+  /// <step_size> If defined this will allow the comms model to run at a
+  /// higher frequency than the physics engine. This is useful when dealing
+  /// with ranging. If the <step_size> is set larger than the physics engine dt
+  /// then the comms model step size will default to dt.
+  /// Note: for consistency it is adviced that the dt is a multiple of timestep.
+  /// Units are in seconds.
+  ///    <messages_topic>: Topic name where the broker receives all the incoming
+  ///                      messages. The default value is "/broker/msgs"
+  ///    <bind_service>: Service name used to bind an address.
+  ///                    The default value is "/broker/bind"
+  ///    <unbind_service>: Service name used to unbind from an address.
+  ///                      The default value is "/broker/unbind"
+  ///
+  /// Here's an example:
+  /// <physics name="1ms" type="ignored">
+  ///   <max_step_size>2</max_step_size>
+  ///   <real_time_factor>1.0</real_time_factor>
+  /// </physics>
+  /// <plugin
+  ///   filename="ignition-gazebo-perfect-comms-system"
+  ///   name="ignition::gazebo::systems::PerfectComms">
+  ///   <step_size>1</step_size>
+  /// </plugin>
+  class ICommsModel
+      : public System,
+        public ISystemConfigure,
+        public ISystemPreUpdate
+  {
+    /// \brief Destructor.
+    public: virtual ~ICommsModel() = default;
+
+    // Documentation inherited.
+    public: void Configure(const Entity &_entity,
+                           const std::shared_ptr<const sdf::Element> &_sdf,
+                           EntityComponentManager &_ecm,
+                           EventManager &_eventMgr) override
+    {
+      sdf::ElementPtr sdfClone = _sdf->Clone();
```

Comment:
Is it right to have the implementations in the .hh file?

Reply:
I just created an `ICommsModel.cpp` in [c56d177](https://github.com/ignitionrobotics/ign-gazebo/pull/1416/commits/c56d1779074de4d2a4e2429a5803110bb1cf7478).